### ISSUE TITLE
Reorganize API documentation

### ISSUE THREAD

**Issue by trevorhardy** [GMLC-TDC/HELICS#934]
As I'm working on forming a proposal to re-organize the documentation, I had a question on how the APIs work across the various language bindings. Specifically, I'd like the new API reference to have one entry for each method showing the exact name used in each binding. This would allow use to have one API reference to describe the functionality of each API (rather than having an API reference where the description of the API is the same and only the name of the method changes). Questions:

1. Is it true that the language bindings only differ by function/method name or are not all language bindings equal and there is functionality that only some support?
2. More generally, do you think this kind of unified API reference is a good idea? If not, any other ideas on how to we can have one place where the description of the API resides while respecting the language specifics for each bound language?

**Comment by kdheepak** [comment]
Can you describe with examples what you mean? I didn't follow what you meant by "one API reference to describe the functionality of each API".

The C interface is documented [here](https://helics.readthedocs.io/en/latest/c-api-reference/index.html). Currently, only the Julia language bindings are documented [here](https://gmlc-tdc.github.io/HELICS.jl/latest/api/). We can / are planning to generate a page similar to Julia's documentation for the Python language binding. Currently, Python users of HELICS are expected to look at either the C API reference or the Julia API reference (or both), to know what functions exist and what their call signature should be. Python users can also type 

```python
import helics
help(helics.helicsFunctionNameHere)
``` 

to get the docstring for the `helics.helicsFunctionNameHere` function. See example below:

```ipython
$ ipython
Python 3.7.3 (default, Mar 27 2019, 16:54:48)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import helics

In [2]: help(helics.helicsGetVersion)
Help on function helicsGetVersion in module helics:

helicsGetVersion() -> 'char const *'
    get a version string for HELICS


In [3]:
```

**Comment by trevorhardy** [comment]
I was thinking something like what I've mocked up below. I don't know if the function/method is called the same thing in each language but I do know that the data types will be different. My idea is trying to avoid the duplication of the (admittedly currently brief) text that is describing what the API does. 

The grand dream is a single, unified, API reference that doesn't take a ton of time to maintain. I don't know enough about how all of this is currently generated, though, to have an understanding of how realistic it is.

`C: HELICS_EXPORT void helicsBrokerAddDestinationFilterToEndpoint(
    helics_broker broker, 
    const char * filter, 
    const char * endpoint, 
    helics_error * err)
C++: HELICS_EXPORT void helicsBrokerAddDestinationFilterToEndpoint | ( 
    helics_broker broker,
    const char * filter,
    const char * endpoint,
    helics_error * err)
Python: ...
MATLAB: ...
Julia: ...


link a named filter to a destination endpoint

Parameters
broker: the broker to generate the connection from
filter: the name of the filter (cannot be NULL)
endpoint: the name of the endpoint to filter the data going to (cannot be NULL)
err: a helics_error object, can be NULL if the errors are to be ignored`

**Comment by trevorhardy** [comment]
Its been a while since I raised this issue (fallen off my radar a bit until lately as we've been planning on reworking the API documentation). 

So, at the risk of repeating myself, how possible is it do to something like I've outlined on my Dec 2, 2019 comment? One entry for each API with a common description of what the API does followed by details of the implementation for each language? I'm interested and willing to work on writing the descriptive language for each API call and I'm assuming that text would reside in a single location. 

Thoughts?

**Comment by trevorhardy** [comment]
Discussion at the developer's meeting today has us leaning towards a few specific steps: 

1. Implement the API documentation for two sets of languages: C++ and everything else (C-based languages)
2. Continue use Doxygen for the C++ documentation.
3. For all the C-based language bindings, use LibClang to generate the documentation.
4. Use LibClang to pull the description of the function from the Doxygen comments and apply it to the C-based API documentation.

**Comment by trevorhardy** [comment]
As an example of why having integrated API documentation would be helpful, I'm working on documentation for the JSON configuration options and would love to link to the appropriate point in the API documentation that corresponds to that particular configuration option. Because we don't have a single API reference, if I want to be comprehensive I'll need to link to multiple API references (C, C++, Python, Julia, etc) for each of these options. If we were able to orient the documentation around the API rather than the language in which it was implemented, I could link to a single page where the user would be able to learn what the API does and then see the details of the call that are pertinent to the language they are using.
